### PR TITLE
Replace raw layer pointers with smart pointers

### PR DESCRIPTION
**Pull request by timmoon10**
We would like to use Cereal to serialize a model's layer graph, but it doesn't support serializing raw pointers. However, it does support serializing `std::weak_ptr`. This PR goes through and replaces all raw layer pointer state with `std::weak_ptr<Layer>`. The only significant substantive functional change is in the unpooling layer (use the hint layer instead of maintaining a pointer to a pooling layer). [Bamboo has no new failures](https://lc.llnl.gov/bamboo/browse/LBANN-TIM308-7).

Note that `std::weak_ptr` points to data in a `std::shared_ptr`. This is suboptimal since we would like for layers to be explicitly owned by models, and hence we would like to keep the layers in `std::unique_ptr<Layer>` s. There is an experimental [`std::observer_ptr`](https://en.cppreference.com/w/cpp/experimental/observer_ptr), but it is not available as of C++20 and [Bjarne doesn't like it](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1408r0.pdf). To make usage clear, I've typedef'ed `OwningLayerPtr=std::shared_ptr<Layer>` and `ViewingLayerPtr=std::weak_ptr<Layer>`.

### CODE REVIEW

File: src/layers/layer.cpp [hunk 1]
```diff
@@ -582,50 +528,189 @@ void Layer::write_proto(lbann_data::Layer* proto) const {
...
-void Layer::add_parent_layer(const Layer* parent) {
-  const auto parent_pos = std::find(m_parent_layers.begin(),
-                                    m_parent_layers.end(),
-                                    parent);
-  if (parent != nullptr
-      && parent != this
-      && parent_pos == m_parent_layers.end()) {
-    m_parent_layers.push_back(parent);
+ViewingLayerPtr Layer::get_parent_layer_pointer(size_t index) const {
+  if (index >= m_child_layers.size()) {
```
Comment (bvanessen) [inline src/layers/layer.cpp:682]: @timmoon10 this should be m_parent_layers.size()

Comment (timmoon10) [reply src/layers/layer.cpp:682]: Good catch.

**Review by bvanessen — APPROVED** [review]
Overall LGTM.  I was able to build locally.  Haven't tested, but since bamboo runs, we should be good.

File: src/proto/factories/layer_graph_factory.cpp [hunk 2]
```diff
@@ -194,17 +150,18 @@ std::vector<std::unique_ptr<Layer>> construct_layer_graph(
...
     // Construct layer
-    std::unique_ptr<Layer> l;
+    OwningLayerPtr l;
 #define TEMPLATE_INSTANTIATION(TensorDataType, T_layout, T_device)      \
     do {                                                                \
       if (proto_datatype == TypeToProtoDataType<TensorDataType>::value  \
           && layout == T_layout                                         \
           && device == T_device) {                                      \
-        l = construct_layer<TensorDataType, T_layout, T_device>(        \
+        auto l_ = construct_layer<TensorDataType, T_layout, T_device>(  \
```
Comment (benson31) [inline src/proto/factories/layer_graph_factory.cpp:159]: `std::shared_ptr<T>` is both assignable and constructible from `std::unique_ptr<T>` rvalues.
```suggestion
        l = construct_layer<TensorDataType, T_layout, T_device>(  \
```

File: src/proto/factories/layer_graph_factory.cpp [hunk 3]
```diff
@@ -194,17 +150,18 @@ std::vector<std::unique_ptr<Layer>> construct_layer_graph(
...
           comm,                                                         \
           training_dr_linearized_data_size,                             \
           num_parallel_readers,                                         \
           proto_layer);                                                 \
+        l.reset(l_.release());                                          \
```
Comment (benson31) [inline src/proto/factories/layer_graph_factory.cpp:164]:
```suggestion
```

**Review by benson31 — APPROVED** [review]
Overall this seems fine. Certainly sufficient to accomplish the goals.

**Comment by timmoon10** [comment]
I'll address the comments in the next PR.

**Pull request merged**